### PR TITLE
Add virtual resource/data packs

### PR DESCRIPTION
**Pull request by Juuxel**
Adds virtual data and resource packs that don't exist as real files, but as a map with input stream providers. They can be registered during mod initialization using `VirtualResourcePackManager.INSTANCE.addPack()`.

To-do:
- [x] `/datapack extractvirtual` for extracting virtual packs
  - An actual client-side command (with [Cotton Client Commands](https://github.com/CottonMC/ClientCommands)\) or a `world.isClient` check?
- [x] Resource pack support
- [x] Test that overriding works
- ~~Support for optional but enabled-by-default resource packs~~ (edit: seems to require a bit more tweaking, not doing that right now)
- [x] Automatic namespace detection

### CODE REVIEW

**Comment by Frontrider** [comment]
One question: in the loading order, is it before or after the user defined ones?

**Comment by Juuxel** [comment]
> in the loading order, is it before or after the user defined ones?

It should be before, but I haven't tested whether it works yet.

File: src/main/java/io/github/cottonmc/cotton/datapack/virtual/VirtualDataPack.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    String getId(int index) {
+        return String.format("virtual/%d_%s", index, id);
+    }
+
+    @Nullable
+    @Override
+    public <T> T parseMetadata(ResourceMetadataReader<T> reader) {
+        return reader.fromJson(DataFixUtils.make(new JsonObject(), object -> {
+            object.addProperty("pack_format", 1);
```
Comment (Reoseah) [inline src/main/java/io/github/cottonmc/cotton/datapack/virtual/VirtualDataPack.java]: shouldn't format be `SharedConstants.getGameVersion().getPackVersion()`?

Comment (Reoseah) [reply src/main/java/io/github/cottonmc/cotton/datapack/virtual/VirtualDataPack.java]: oh, I'm just not viewing the last version

**Comment by Frontrider** [comment]
I'm gonna be a bit nitpicky. One could generate images that can be provided by the virtual resource packs. I think there are probably better ways when you do dynamic stuff anyways, but it should still be kept in mind.

**Comment by Juuxel** [comment]
From the discord chat:

For images in virtual resource packs, I can replace the string suppliers with some `InputStreamProvider` that can be constructed through an output stream or a string supplier.

So `InputStreamProvider.of(() -> "hello")` or `InputStreamProvider.ofOutput(stream -> <write your things to the output stream>)`.

**Comment by Frontrider** [comment]
That should work.

**Comment by Frontrider** [comment]
I added a command called `exportvirtual` that takes in an argument that is a type of resource (`assets` or `data`), and then prints out all of the virtual resources of that type. We can also add the exact name of the pack as a third optional argument.

**Comment by Frontrider** [comment]
world.isClient is out of question, because the command will only run on a server.

**Comment by Juuxel** [comment]
Thanks for the command, @Frontrider.

> world.isClient is out of question, because the command will only run on a server.
Ah, of course. Do we want to make an optional dependency on Cotton Client Commands (it uses an entrypoint/plugin system for registering)?

**Comment by Frontrider** [comment]
We can make that an optional dependency.

**Pull request merged**